### PR TITLE
Add a basic material slider

### PR DESCRIPTION
**Pull request by abarth**
The slider in this patch works, but it's missing animations and many subtle
touches.

### CODE REVIEW

**Comment by abarth** [comment]
@HansMuller @Hixie

File: examples/material_gallery/lib/slider_demo.dart [hunk 1]
```diff
@@ -0,0 +1,56 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/material.dart';
+
+import 'widget_demo.dart';
+
```
Comment (Hixie) [inline examples/material_gallery/lib/slider_demo.dart]: why not merge the SliderExample and the SliderDemo?
At least make SliderDemo Stateless if you want to keep it.

File: packages/flutter/lib/src/material/slider.dart [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  Widget build(BuildContext context) {
+    return new _SliderWrapper(
+      value: value,
+      primaryColor: Theme.of(context).accentColor,
+      onChanged: onChanged
+    );
+  }
+}
+
+class _SliderWrapper extends LeafRenderObjectWidget {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/slider.dart]: _SliderRenderObjectWidget

File: packages/flutter/lib/src/material/slider.dart [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    _drag = new HorizontalDragGestureRecognizer(router: FlutterBinding.instance.pointerRouter)
+      ..onStart = _handleDragStart
+      ..onUpdate = _handleDragUpdate
+      ..onEnd = _handleDragEnd;
+  }
+
+  double get value => _value;
+  double _value;
+  void set value(double newValue) {
+    if (newValue == _value)
```
Comment (Hixie) [inline packages/flutter/lib/src/material/slider.dart]: assert that it's in the range 0.0..1.0.

File: packages/flutter/lib/src/material/slider.dart [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+class _RenderSlider extends RenderConstrainedBox {
+  _RenderSlider({
+    double value,
+    Color primaryColor,
+    this.onChanged
+  }) : _value = value,
+       _primaryColor = primaryColor,
+        super(additionalConstraints: const BoxConstraints.tightFor(width: 144.0, height: 2 * _kReactionRadius)) {
+    _drag = new HorizontalDragGestureRecognizer(router: FlutterBinding.instance.pointerRouter)
```
Comment (Hixie) [inline packages/flutter/lib/src/material/slider.dart]: assert that value is non-null and in the range 0.0..1.0.

File: packages/flutter/lib/src/material/slider.dart [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+  ValueChanged<double> onChanged;
+
+  HorizontalDragGestureRecognizer _drag;
+  bool _active = false;
+  double _currentDragValue = 0.0;
+
+  void _handleDragStart(Point globalPosition) {
+    _active = true;
+    markNeedsPaint();
+    if (onChanged != null) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/slider.dart]: you go active regardless of whether you have an onChanged, but you don't change unless you do. It seems like these should be consistent. Either we set _currentDragValue in all cases, or we use the lack of onChanged to disable ourselves entirely.

File: packages/flutter/lib/src/material/slider.dart [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    _active = false;
+    _currentDragValue = 0.0;
+    markNeedsPaint();
+  }
+
+  bool hitTestSelf(Point position) => true;
+
+  void handleEvent(InputEvent event, BoxHitTestEntry entry) {
+    if (event.type == 'pointerdown')
+      _drag.addPointer(event);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/slider.dart]: only do this if we're enabled? (i.e. have an onChanged)

File: packages/flutter/lib/src/material/slider.dart [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+
+    double trackCenter = offset.dy + size.height / 2.0;
+    double trackLeft = offset.dx;
+    double trackTop = trackCenter - 1.0;
+    double trackBottom = trackCenter + 1.0;
+    double trackRight = offset.dx + size.width;
+    double trackActive = offset.dx + size.width * value;
+
+    Paint primaryPaint = new Paint()..color = _primaryColor;
+    Paint trackPaint = new Paint()..color = _active ? _kActiveTrackColor : _kInactiveTrackColor;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/slider.dart]: is there a "disabled" color?
should these colours be overridable?

File: packages/flutter/lib/src/material/slider.dart [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+    canvas.drawRect(new Rect.fromLTRB(trackLeft, trackTop, trackRight, trackBottom), trackPaint);
+    canvas.drawRect(new Rect.fromLTRB(trackLeft, trackTop, trackActive, trackBottom), primaryPaint);
+
+    Point activeLocation = new Point(trackActive, trackCenter);
+    if (_active) {
+      Paint reactionPaint = new Paint()..color = _primaryColor.withAlpha(0x33);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/slider.dart]: 0x33 should be a constant somewhere

File: packages/flutter/lib/src/material/slider.dart [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  void handleEvent(InputEvent event, BoxHitTestEntry entry) {
+    if (event.type == 'pointerdown')
+      _drag.addPointer(event);
+  }
+
+  void paint(PaintingContext context, Offset offset) {
+    final Canvas canvas = context.canvas;
+
+    double trackCenter = offset.dy + size.height / 2.0;
+    double trackLeft = offset.dx;
```
Comment (Hixie) [inline packages/flutter/lib/src/material/slider.dart]: you need some padding on this guy or you won't be able to hit the left side of the knob when it's on the left and the right side of the knob when it's on the right.

I guess you need to add 2*kReactionRadius to the width, then paint at 1*kReactionRadius in from the offset.dx, or some such.

File: packages/flutter/lib/src/material/slider.dart [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+final Color _kActiveTrackColor = Colors.grey[500];
+
+class _RenderSlider extends RenderConstrainedBox {
+  _RenderSlider({
+    double value,
+    Color primaryColor,
+    this.onChanged
+  }) : _value = value,
+       _primaryColor = primaryColor,
+        super(additionalConstraints: const BoxConstraints.tightFor(width: 144.0, height: 2 * _kReactionRadius)) {
```
Comment (Hixie) [inline packages/flutter/lib/src/material/slider.dart]: 144.0 should be a constant somewhere

**Comment by Hixie** [comment]
LGTM with minor comments

File: examples/material_gallery/lib/slider_demo.dart (hunk 1 above)
Comment (abarth) [inline examples/material_gallery/lib/slider_demo.dart]: I was going to make more examples, but you're right that it doesn't make sense currently.  I'll merge it back in.

File: packages/flutter/lib/src/material/slider.dart (hunk 7 above)
Comment (abarth) [inline packages/flutter/lib/src/material/slider.dart]: The colors are in the spec.  The disabled color is the same as the inactive color.

File: packages/flutter/lib/src/material/slider.dart (hunk 9 above)
Comment (abarth) [inline packages/flutter/lib/src/material/slider.dart]: Yeah.

**Pull request merged**